File: EventQRBackend/eventqr/src/main/java/com/thedavelopers/eventqr/features/reports/model/dto/EventReportDtos.java
package com.thedavelopers.eventqr.features.reports.model.dto;

import java.time.Instant;
import java.time.LocalDate;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import com.thedavelopers.eventqr.features.reports.model.ReportEmptyState;
import com.thedavelopers.eventqr.features.reports.model.ReportFilterStatus;
import com.thedavelopers.eventqr.features.reports.model.ReportType;

public final class EventReportDtos {

    private EventReportDtos() {
    }

    public record EventReportSummaryResponse(UUID eventId, String eventName, long registeredCount, long checkedInCount,
                                             long exitedCount, boolean hasAnyRecords) {
    }

    public record EventReportFilters(LocalDate startDate, LocalDate endDate, String attendeeQuery,
                                     ReportFilterStatus status) {
    }

    public record EventReportRow(List<String> values) {
    }

    public record EventReportResponse(UUID eventId, ReportType reportType, String reportTitle, String eventName,
                                      Instant generatedAt, List<String> columns, List<EventReportRow> rows,
                                      Map<String, Long> chartSeries, ReportEmptyState emptyState,
                                      EventReportFilters appliedFilters) {
    }

    public record ReportExportRequest(String format, LocalDate startDate, LocalDate endDate, String attendeeQuery,
                                      ReportFilterStatus status) {
    }
}

File: EventQRMobile/app/src/main/java/com/thedavelopers/eventqr/features/reports/model/dto/EventReportDtos.kt
package com.thedavelopers.eventqr.features.reports.model.dto

import java.io.Serializable
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale

data class EventReportSnapshot(
    val totalAttendees: Int = 0,
    val registeredCount: Int = 0,
    val enteredCount: Int = 0,
    val exitedCount: Int = 0,
    val noShowCount: Int = 0,
    val attendanceCount: Int = 0,
    val claimsCount: Int = 0,
    val boothSessionVisits: Int = 0,
    val rewardsRedeemed: Int = 0,
    val totalPointsEarned: Int = 0,
    val approvedTransactions: Int = 0,
    val rejectedTransactions: Int = 0,
)

enum class EventReportType : Serializable {
    ROSTER,
    NO_SHOWS,
    ENTRY_LOGS,
    ATTENDANCE,
    CLAIMS,
    BOOTH_VISITS,
    EXIT_LOGS,
    POINTS,
}

enum class EventReportFilterStatus : Serializable {
    ALL,
    APPROVED,
    REJECTED,
}

enum class EventReportEmptyState : Serializable {
    NONE,
    NO_FILTER_MATCH,
    NO_EVENT_RECORDS,
}

data class EventReportSummaryDto(
    val eventId: String = "",
    val eventName: String = "",
    val registeredCount: Int = 0,
    val checkedInCount: Int = 0,
    val exitedCount: Int = 0,
    val hasAnyRecords: Boolean = false,
) : Serializable

data class EventReportFiltersDto(
    val startDate: LocalDate? = null,
    val endDate: LocalDate? = null,
    val attendeeQuery: String? = null,
    val status: EventReportFilterStatus = EventReportFilterStatus.ALL,
) : Serializable

data class EventReportRowDto(
    val values: List<String> = emptyList(),
) : Serializable

data class EventReportDto(
    val eventId: String = "",
    val reportType: EventReportType = EventReportType.ROSTER,
    val reportTitle: String = "",
    val eventName: String = "",
    val generatedAt: String? = null,
    val columns: List<String> = emptyList(),
    val rows: List<EventReportRowDto> = emptyList(),
    val chartSeries: Map<String, Long> = emptyMap(),
    val emptyState: EventReportEmptyState = EventReportEmptyState.NONE,
    val appliedFilters: EventReportFiltersDto = EventReportFiltersDto(),
) : Serializable {

    val generatedAtInstant: Instant?
        get() = generatedAt?.let { Instant.parse(it) }
}

data class EventReportExportRequestDto(
    val format: String = "CSV",
    val startDate: LocalDate? = null,
    val endDate: LocalDate? = null,
    val attendeeQuery: String? = null,
    val status: EventReportFilterStatus = EventReportFilterStatus.ALL,
)

data class EventReportCatalogItem(
    val reportType: EventReportType,
    val label: String,
    val iconRes: Int,
    val iconTint: Int,
    val iconBg: Int,
) : Serializable
